text/0000-stabilize-marker-freeze.md

```diff
@@ -0,0 +1,66 @@
+- Feature Name: `stabilize_marker_freeze`
+- Start Date: 2024-05-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Stabilize `core::marker::Freeze` in trait bounds.
+
+# Motivation
+[motivation]: #motivation
+
+In some contexts, guaranteeing that a value cannot be modified is a requirement to a construct behaving properly (`const` references, keys of a structured map...).
+
+While this looks achievable by only exposing immutable references, this is actually insufficient due to the (necessary) existence of interior mutability.
+
+Internally, the compiler uses `core::marker::Freeze` to identify types that are known not to have interior mutability.
+
+This RFC seeks to stabilize this trait for trait bounds for the following reasons:
+- As explained above, some existing constructs are left to "trust" that the values they'd need to freeze won't be mutated through interior mutability. This is a potential bug source for these constructs.
+- With [this PR](https://github.com/rust-lang/rust/issues/121250), a breaking change was introduced (and stabilized in 1.78):
+	- This change prevents associated consts from containing references to values that don't implement `core::marker::Freeze`. Since this bound cannot be added to generics, this prevents associated consts from containing references to generics altogether.
+	- This change was done in order to prevent potential unsoundness, but it also completely prevents sound uses, such as the ones reported in [this regression issue](https://github.com/rust-lang/rust/issues/123281).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC simply seeks to act as the stabilization RFC for `core::marker::Freeze` in trait bounds.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The work necessary for this RFC has already been done and merged in [this PR](https://github.com/rust-lang/rust/issues/121675), and a [tracking issue](https://github.com/rust-lang/rust/issues/121675) was opened.
+
+However, it was deemed that adding an Auto Trait to Rust's public API should go through the proper RFC process, but the RFC was never created.
```

Comment:
The RFC should spell out the drawbacks of a new auto trait more explicitly.

Reply:
Specifically, auto traits are a semver hazard: adding a new private field to a type that already has private fields is a breaking change if that field removes the `Freeze` property of the surrounding type.

However, at least under some conditions this is already a breaking change today: if there are any constants `C` of a type `T`, then `const { &C }` will compile if `T: Freeze`, but may fail to compile otherwise. By exposing `Freeze` bounds it is now always a breaking change, even if there are no constants.

The [semver guide](https://doc.rust-lang.org/cargo/reference/semver.html) doesn't seem to discuss auto traits at all, or did I miss the relevant section @ehuss?

Reply:
It hasn't been added yet. Many missing items are tracked in https://github.com/rust-lang/cargo/issues/8736, including "Adding or removing private fields of a struct can be a breaking change if those fields change the auto traits.". I'm sure there are many more.